### PR TITLE
UX tweaks & refactoring

### PR DESCRIPTION
**Pull request by andre1050**
Includes a variety of UI/UX tweaks/improvements from feedback last week, as well as a few others I've wanted to make once we've agreed on some basic UX patterns.

- Redirect to new test on creating one
- Redirect to next test in list when deleting one
- Bypass save/discard confirmations when no changes
- Persist test name/description/status in test list until changes have been saved
- Persist test title in section heading until changes saved
- Display "modified" label in test list when that test has unsaved changes
- Reset active tab when navigating between tests
- Display test creation date/author when available
- Show a "nice" empty list placeholder when no tests in the slot

Aside from these, also refactored most components to abstract things like colours, helper functions and list items. The basic principle remains that we have two copies of the `test` object: one is a working copy and always reflects the state of the object/array at a particular point (order in the array, and data in the object properties), and separately a clone with remains unchanged until the user explicitly saves the changes, at which point the working copy replaces the clone/original. This makes it easy to compare the objects to detect changes, as well as reverting to previous version without having to track individual changes.

### CODE REVIEW

File: ui/.env [hunk 1]
```diff
@@ -0,0 +1 @@
+REACT_APP_AUTOMAT_API_URL="http://localhost:9000"
```
Comment (andre1050) [inline ui/.env:1]: I should have added this file to `.gitignore` but will update the PR before merging.

File: ui/src/utils/patchTests.ts [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+  return fetch(apiUrl, {
+    method: 'PATCH',
+    headers: {
+      'Content-Type': 'application/json',
+    },
+    body: JSON.stringify({ tests: tests }),
+  })
+    .then(checkForErrors)
+    .then((response) => response.json());
+};
```
Comment (tjmw) [inline ui/src/utils/patchTests.ts:16]: I wonder if defining all operations against the API in a single client would make sense? I think there will be a bunch of common building blocks that all operations have in common. What do you think?

File: ui/src/utils/patchTests.ts [hunk 3]
```diff
@@ -0,0 +1,16 @@
+import { Test } from '../types';
+import { checkForErrors } from '../utils/checkForErrors';
+
+export const patchTestsBySlot = (slotId: string, tests: Test[]) => {
+  const apiPrefix = process.env.REACT_APP_AUTOMAT_API_URL;
+  const apiUrl = `${apiPrefix}/admin/slotss/${slotId}`;
```
Comment (tjmw) [inline ui/src/utils/patchTests.ts:6]: Tiny typo:

```suggestion
  const apiUrl = `${apiPrefix}/admin/slots/${slotId}`;
```

File: ui/src/App.tsx [hunk 4]
```diff
@@ -1,16 +1,24 @@
...
 import { Spinner } from './components/Spinner';
-import { FeedbackMessage } from './components/FeedbackMessage';
+import { Notification } from './components/Notification';
+import { Slot } from './types';
 
 import { Shell } from './Shell';
-import { Slots } from './screens/Slots';
-import { Tests } from './screens/Tests';
+import { Slots as SlotsScreen } from './screens/Slots';
+import { Tests as TestsScreen } from './screens/Tests';
```
Comment (tjmw) [inline ui/src/App.tsx:11]: Would it be worth defining a `./screens/index.ts` so we can collapse the imports into one like:

```
import { Tests as TestsScreen, Slots as SlotsScreen } from './screens'
```

What do you think?

File: ui/.env (hunk 1 above)
Comment (andre1050) [reply ui/.env]: File deleted.

File: ui/src/utils/patchTests.ts (hunk 3 above)
Comment (andre1050) [reply ui/src/utils/patchTests.ts]: Thanks! 🙈

File: ui/src/App.tsx (hunk 4 above)
Comment (andre1050) [reply ui/src/App.tsx:11]: Good question. My feeling is that doing that is just as worth it as adding an index file that imports/exports all the components under `./components` and I don't think that's a very common pattern in TS projects 🤔 But I could be wrong. I propose we rename these two "screen components" explicitly to `SlotsScreen` and `TestsScreen` so at least we can do a simple named import without having to rename things.

If we do get to a point where we have many screen/page components, we may want to revisit this and add the index file you suggest, to shorten the imports block, but I don't think it's worth doing yet.

File: ui/src/utils/patchTests.ts (hunk 2 above)
Comment (andre1050) [reply ui/src/utils/patchTests.ts:16]: Yeah, I think it does. No need for each of these functions to live in a separate file at least. I'll follow up with a separate PR to capture the variants/filter work as well.

**Pull request merged**